From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Avnish Chouhan <avnish@linux.ibm.com>
Date: Thu, 25 Jun 2026 18:23:36 +0200
Subject: [PATCH] ieee1275/powerpc: implements fibre channel discovery for
 ofpathname

grub-ofpathname doesn't work with fibre channel because there is no
function currently implemented for it.
This patch enables it by prividing a function that looks for the port
name, building the entire path for OF devices.

Signed-off-by: Diego Domingos <diegodo@br.ibm.com>
Signed-off-by: Avnish Chouhan <avnish@linux.ibm.com>
Reviewed-by: Leo Sandoval <lsandova@redhat.com>
Reviewed-by: Sudhakar Kuppusamy <sudhakar@linux.ibm.com>
Part-of: <https://gitlab.freedesktop.org/gnu-grub/grub/-/merge_requests/146>
---
 grub-core/osdep/linux/ofpath.c | 58 +++++++++++++++++++++---------------------
 1 file changed, 29 insertions(+), 29 deletions(-)

diff --git a/grub-core/osdep/linux/ofpath.c b/grub-core/osdep/linux/ofpath.c
index 72772c87a66f..53cf39cd176d 100644
--- a/grub-core/osdep/linux/ofpath.c
+++ b/grub-core/osdep/linux/ofpath.c
@@ -351,36 +351,36 @@ of_path_of_ide(const char *sys_devname __attribute__((unused)), const char *devi
   return ret;
 }
 
-
+#define BNAME_SIZE 150
 static void
-of_fc_port_name(const char *path, const char *subpath, char *port_name)
+of_fc_port_name (const char *path, const char *subpath, char *port_name)
 {
   char *bname, *basepath, *p;
   int fd;
 
-  bname = xmalloc(sizeof(char)*150);
-  basepath = xmalloc(strlen(path));
+  bname = xmalloc (sizeof (char) * BNAME_SIZE);
+  basepath = xmalloc (strlen (path));
 
   /* Generate the path to get port name information from the drive */
-  strncpy(basepath,path,subpath-path);
-  basepath[subpath-path-1] = '\0';
-  p = get_basename(basepath);
-  snprintf(bname,sizeof(char)*150,"%s/fc_transport/%s/port_name",basepath,p);
+  strncpy (basepath, path, subpath-path);
+  basepath[subpath-path - 1] = '\0';
+  p = get_basename (basepath);
+  snprintf (bname, sizeof (char) * BNAME_SIZE, "%s/fc_transport/%s/port_name", basepath, p);
 
   /* Read the information from the port name */
   fd = open (bname, O_RDONLY);
   if (fd < 0)
     grub_util_error (_("cannot open `%s': %s"), bname, strerror (errno));
 
-  if (read(fd,port_name,sizeof(char)*19) < 0)
+  if (read (fd, port_name, sizeof (char) *19) < 0)
     grub_util_error (_("cannot read `%s': %s"), bname, strerror (errno));
 
-  sscanf(port_name,"0x%s",port_name);
-  
-  close(fd);
+  sscanf (port_name, "0x%s", port_name);
 
-  free(bname);
-  free(basepath);
+  close (fd);
+
+  free (bname);
+  free (basepath);
 }
 
 #ifdef __sparc__
@@ -610,16 +610,16 @@ of_path_of_scsi(const char *sys_devname __attribute__((unused)), const char *dev
   digit_string = trailing_digits (device);
   if (strncmp (of_path, "/vdevice/", sizeof ("/vdevice/") - 1) == 0)
     {
-      if(strstr(of_path,"vfc-client"))
-      {
-	char * port_name = xmalloc(sizeof(char)*17);
-	of_fc_port_name(sysfs_path, p, port_name);
-	
-	snprintf(disk,sizeof(disk),"/%s@%s", disk_name, port_name);
-	free(port_name);
-      }
+      if (strstr (of_path, "vfc-client"))
+        {
+	   char * port_name = xmalloc (sizeof (char) * 17);
+	   of_fc_port_name (sysfs_path, p, port_name);
+
+          snprintf (disk, sizeof (disk), "/%s@%s", disk_name, port_name);
+	   free (port_name);
+        }
       else
-      {
+        {
       unsigned long id = 0x8000 | (tgt << 8) | (bus << 5) | lun;
       if (*digit_string == '\0')
 	{
@@ -634,12 +634,12 @@ of_path_of_scsi(const char *sys_devname __attribute__((unused)), const char *dev
 		   "/%s@%04lx000000000000:%c", disk_name, id, 'a' + (part - 1));
 	}
 	}
-    } else if (strstr(of_path,"fibre-channel")||(strstr(of_path,"vfc-client"))){
-	char * port_name = xmalloc(sizeof(char)*17);
-	of_fc_port_name(sysfs_path, p, port_name);
-	
-	snprintf(disk,sizeof(disk),"/%s@%s", disk_name, port_name);
-	free(port_name);
+    } else if (strstr (of_path, "fibre-channel") || (strstr (of_path, "vfc-client"))){
+        char * port_name = xmalloc (sizeof (char) * 17);
+        of_fc_port_name (sysfs_path, p, port_name);
+
+	 snprintf (disk, sizeof (disk), "/%s@%s", disk_name, port_name);
+	 free (port_name);
     }
   else
     {
